### PR TITLE
Unix Viewers - 2nd attempt 

### LINKED ISSUE #5945: Image.open on Ubuntu 20.04 gives the error "No images found" in the viewer
The viewer is opened, but a "No images found" error is displayed.  Pillow version `pillow-9.0.0`.  This may be the same bug: https://stackoverflow.com/questions/63579415/why-does-pil-fail-to-display-images-with-error-no-images-found-in-file-tmp

Code to reproduce:

```python
from PIL import Image

im = Image.open(r"/var/lib/app-info/icons/ubuntu-focal-updates-universe/48x48/inkscape_inkscape.png")
im.show()
```

Changing the code in the `UnixViewer` class in `ImageShow.py` seems to fix it.  I'm not sure why the current code is writing the filename to a file and then reading it back from stdin, using `shell=True` in `Popen`.   Deleting the `rm -f $im"` command in the `Popen` call seems to fix the problem.  It also seems to work using `cat | xargs -L1` in the Popen command instead of setting the shell variable `im`.

This simpler code replacement also seems to work, without setting `shell=True`:

```python
def show_file(self, file, **options):
    """Display given file"""
    command = self.get_command_ex(file, **options)[0]
    subprocess.Popen([command, file])
    return 1
```

### PR DESCRIPTION
**Pull request by alv2017**
Fixes #5945, #5976

Changes proposed in this pull request:

 *  Viewer class properties **format** and **options** have been moved to the parent Viewer class, those properties were removed from the Viewer child classed. **format** is set to **"PNG"** and **options** are set to **{"compress_level": 1}** The exception is IPythonViewer class, for this class **format** was set to **None** and options are set to **{}**.
* Python threads were used to control opening/closing of images by external Linux/Unix viewers.
* Within current setup, when working with Linux viewers the Python program remains opened till the user closes the image in the viewer.
* Implementation description: there are viewer thread and monitoring thread. Viewer thread is responsible for opening an image in a viewer; the monitoring thread is responsible for removal of temporary images after user closed the image in the viewer.
* Corresponding documentation is updated.
* Affected tests are updated: Tests/test_imageshow.py, Tests/helper.py

### CODE REVIEW

**Comment by alv2017** [comment]
I ran all the workflows in my repository, all the workflows except Windows passed.

**Comment by hugovk** [comment]
The Windows failure was a bug in `actions/setup-python`, now fixed: https://github.com/actions/setup-python/issues/328. I've restarted the Windows jobs.

**Comment by alv2017** [comment]
It failed on  Python pypy-3.7 x64:
Process completed with exit code -1073741819.
No files were found with the provided path: Tests/errors. No artifacts will be uploaded.

**Comment by radarhere** [comment]
I reran the Windows jobs again, and they have all passed now.

**Comment by alv2017** [comment]
@hugovk, @radarhere 

Please let me know if this PR has a chance of being accepted. Thanks!

File: Tests/test_imageshow.py [hunk 1]
```diff
@@ -41,9 +41,11 @@ def show_image(self, image, **options):
     ImageShow._viewers.pop(0)
 
 
-@pytest.mark.skipif(
-    not on_ci() or is_win32(),
-    reason="Only run on CIs; hangs on Windows CIs",
+@pytest.mark.skip(
+    reason="""Current implementation of some viewers requires manual closing of an image,
+        because of that the tests calling show() method will hang infinitely.
+        Please also note that this test duplicates test_viewer_show() test.
```
Comment (radarhere) [inline Tests/test_imageshow.py:47]: I don't think it is a duplicate. `test_viewer_show()` registers a dummy viewer, to check that `Viewer.show()` calls `show_image()`. `test_show()`  runs an actual viewer.

**Comment by alv2017** [comment]
@radarhere I corrected the test skip message.

File: src/PIL/ImageShow.py [hunk 2]
```diff
@@ -234,150 +235,80 @@ def show_file(self, path=None, **options):
...
+
+        path = quote(path)
+        command = self.get_command(path, **options)
+
+        kwargs = {
+            "stdout": subprocess.PIPE,
+        }
+
+        th = threading.Thread(
+            target=subprocess.run, args=(command.split(),), kwargs=kwargs, name=path
```
Comment (radarhere) [inline src/PIL/ImageShow.py:247]: We've recently had to address the possibility of [the file path having a space in it](https://github.com/python-pillow/Pillow/pull/3450#issuecomment-1009608418). This would `split()` incorrectly in such a situation, yes?

File: src/PIL/ImageShow.py [hunk 3]
```diff
@@ -234,150 +235,80 @@ def show_file(self, path=None, **options):
...
+        return command, executable
 
 
 class GmDisplayViewer(UnixViewer):
     """The GraphicsMagick ``gm display`` command."""
 
     def get_command_ex(self, file, **options):
         executable = "gm"
-        command = "gm display"
+        command = f"gm display {quote(file)}"
```
Comment (radarhere) [inline src/PIL/ImageShow.py:285]: As I've said before, I consider this to be a backwards incompatible change.

File: Tests/helper.py [hunk 4]
```diff
@@ -27,8 +27,7 @@
     class test_image_results:
         @staticmethod
         def upload(a, b):
-            a.show()
-            b.show()
```
Comment (radarhere) [inline Tests/helper.py:31]: What is the reason for this change?

Comment (alv2017) [reply Tests/helper.py:31]: If it is hit, the tests will hang infinitely. I do not see any good reason for a.show(), b.show(): we are interested in return value of the **upload** function, and return value here is None.

File: src/PIL/ImageShow.py (hunk 3 above)
Comment (alv2017) [reply src/PIL/ImageShow.py:285]: I think that it is more convenient to keep a viewer opening shell command at one place. As excuse I will add, that get_command_ex() is a helper function, and the interface of the function was preserved.

File: src/PIL/ImageShow.py (hunk 2 above)
Comment (alv2017) [reply src/PIL/ImageShow.py:247]: 1) I do not think that there is a a split problem, because the file(path) is quoted, and the path always be represented as a single entity.
2) I wouldn't be concerned about users having a temporary path "foo bar baz", I think that they will be able to sort this out.

Comment (alv2017) [reply src/PIL/ImageShow.py]: The solution was to run the viewer command as a shell command. User input has been sanitized in **get_command_ex** stage.  As a result (a) path problem solved; (b) possible security problem solved;
```python
        kwargs = {"shell": True, "stdout": subprocess.PIPE, "stderr": subprocess.PIPE}

        th = threading.Thread(
            target=subprocess.run, args=(command,), kwargs=kwargs, name=path
        )
```

File: Tests/helper.py (hunk 4 above)
Comment (nulano) [reply Tests/helper.py:31]: Huh? This is only used if there is an environment variale `SHOW_ERRORS` set, which means the user is asking to be shown failed image comparisons. Removing this breaks that functionality. Also, the result of the upload function is irrelevant as there is a check in place to see if it is defined. If it isn't defined, it is never called anyway, so if this were to be removed, you would remove the whole if block.

**Comment by radarhere** [comment]
I've created #6039 as an alternate version of this. The primary difference is that it doesn't cause Python to hang while waiting for the image viewer to close. See what you think.

**Comment by hugovk** [comment]
https://github.com/python-pillow/Pillow/pull/6045 has been merged instead. Thank you for a way for the suggestion!
